File: mk_dockerfile.py
import os, sys
import configparser  # use ConfigParser for v2 (code must be modified)

def mk_docker_file(config):
    for section in config.sections():
        # read template file name
        template_file_name = config[section]["template"]
        template_file_name = "template/" + template_file_name

        # replace all "$tag" type tags to matching config values
        data = ""
        with open(template_file_name, "r") as f:
            # replace contents
            data = f.read()
            for key in config[section]:
                if key == "template":
                    continue
                val = config[section][key]
                src_substr = "\"$" + key + "\""
                data = data.replace(src_substr, val)

            # write replaced content to new dockerfile
            # filename = Dockerfile.section_name
            docker_file_name = "dockerfile/" + "Dockerfile" + "." + section
            with open(docker_file_name, "w+") as df:
                wlen = df.write(data)
                if wlen == len(data):
                    print("Done : %s" %(docker_file_name))
                else:
                    print("Fail : %s (%d) != (%d)" %(docker_file_name, wlen, len(data)))


def read_config(config_file):
    config = configparser.ConfigParser()
    config.read(config_file)
    return config

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage : %s config_file" %(os.path.basename(sys.argv[0])))
        sys.exit(1)

    config = read_config(sys.argv[1])
    mk_docker_file(config)

File: withgo/mk_dockerfile.go
package main

import (
	"errors"
	"fmt"
	"github.com/spf13/viper"
	"io/ioutil"
	"log"
	"os"
	"path"
	"strings"
)

var (
	templateDir = "./template/"
	dockerfilePrefix = "Dockerfile"
	dockerfileDir = "./dockerfile/"
	prefix = "\"$"
	postfix = "\""
)

func fatalf(err error) {
	if err != nil {
		log.Panicf("Fatal error config file: %s \n", err)
	}
}

func readConfig(filename string) *viper.Viper {
	v := viper.New()

	v.SetConfigName(filename)
	v.SetConfigType("yaml")
	v.AddConfigPath("./")

	err := v.ReadInConfig()
	fatalf(err)

	return v
}

// Construct map for section name and each sections' item list
func convertToConfigMap(v *viper.Viper) (map[string][]string) {
	keys := v.AllKeys()
	configMap := make(map[string][]string)
	var sectionName string
	var itemName string

	for _, key := range keys {
		s := strings.Split(key, ".")
		sectionName = s[0]
		itemName = s[1]
		configMap[sectionName] = append(configMap[sectionName], itemName)
	}
	return configMap
}

func readTemplate(tfile string) (string, error) {
	var buf []byte
	buf, err := ioutil.ReadFile(tfile)
	if err != nil {
		return "", err
	}

	return string(buf), nil
}

func validItemConfig(sectionName string, itemList []string) error {
	exist := false
	for _, item := range itemList {
		if item == "template" {
			exist = true
			break
		}
	}
	if !exist {
		return errors.New(fmt.Sprintf("template item doesn't exist in [%s] section", sectionName))
	}
	return nil
}

func makeDockerFile(v *viper.Viper) {

	configMap := convertToConfigMap(v)

	for sectionName, itemList := range configMap {
		fatalf( validItemConfig(sectionName, itemList) )

		vkey := sectionName + "." + "template"
		templateFileName := fmt.Sprintf("%v", v.Get(vkey))
		content, err := readTemplate(templateDir + templateFileName)
		fatalf(err)

		for _, item := range itemList {
			if item == "template" {
				continue
			}

			templateStr := prefix + item + postfix
			valueStr := fmt.Sprintf("%v", v.Get(sectionName + "." + item))

			content = strings.ReplaceAll(content, templateStr, valueStr)
		}

		outputFileName := dockerfileDir + dockerfilePrefix + "." + sectionName
		fatalf( ioutil.WriteFile(outputFileName, []byte(content), 0644) )
		fmt.Printf("Done : %s\n", outputFileName)
	}
}

func main() {
	if len(os.Args) < 2 {
		fmt.Printf("Usage : go run %s.go config_file_name\n", path.Base(os.Args[0]))
		os.Exit(0)
	}

	configFile := os.Args[1]

	v := readConfig(configFile)
	makeDockerFile(v)
}
